Ensure adding a broken tar doesn't silently fail

aidanhs opened the pull request:
The change to `daemon/daemonbuilder/builder.go` ensures that if `UntarPath` returns an error, that error will be passed upwards (previously it would log an error at debug, i.e. nothing for most users, and silently succeed). I have deliberately opted to include `io.EOF`, because that shouldn't be received at this level of code.

The change to `pkg/archive/archive.go` is just a simplification.

aidanhs committed:
Ensure adding a broken tar doesn't silently fail

Signed-off-by: Aidan Hobson Sayers <aidanhs@cantab.net>

GordonTheTurtle commented:
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "aphs-fail-on-broken-tar" git@github.com:aidanhs/docker.git somewhere
$ cd somewhere
$ git commit --amend -s --no-edit
$ git push -f
~~~

Ammending updates the existing PR. You **DO NOT** need to open a new one.

calavera commented:
please, fix the formatting.

aidanhs commented:
Sorry! Done.

aidanhs commented:
I'm looking at the test failures...

aidanhs commented:
Ok, I think I've fixed the tests now and added some new ones. `TestBuildAddNonTar` ensures that the existing behaviour is unchanged, `TestBuildAddBrokenTar` tests the new functionality.

As miscellaneous cleanup I moved the tar header size constant inside the archive package, simplified the return at the end of `UntarPath` and make the comment for `IsArchive` more specific.

crosbymichael commented on daemon/daemonbuilder/builder.go:
```diff
@@ -174,11 +174,11 @@ func (d Docker) Copy(c *daemon.Container, destPath string, src builder.FileInfo,
-			return nil
-		} else if err != io.EOF {
-			logrus.Debugf("Couldn't untar to %s: %v", tarDest, err)
+		err := d.Archiver.UntarPath(srcPath, tarDest)
```
can you explain why we are ignoring the EOF here ?

aidanhs replied:
Before it was being ignored, now it's not.

Previously, any error part way during extraction would fall through to adding the file. The only thing that `io.EOF` controlled was a log line that virtually nobody would see.
I've removed that conditional because an `io.EOF` (or any other error) bubbling up to this point means something *has* gone wrong during the untar (e.g. the file is too short to peek the magic bytes from) and so we should error.

crosbymichael replied:
ok, that makes sense, thanks

crosbymichael commented:
LGTM

tiborvass commented on integration-cli/docker_cli_build_test.go:
```diff
@@ -4217,6 +4217,79 @@ RUN cat /existing-directory-trailing-slash/test/foo | grep Hi`
+	}
+}
+
+func (s *DockerSuite) TestBuildAddNonTar(c *check.C) {
```
@aidanhs is this test covering changes in the PR or covering existing code in master?
IOW, should this test pass or fail if tested on master?

TestBuildAddBrokenTar fails on master (which is expected), but TestBuildAddNonTar passes, and I'm not exactly sure if that's expected or not.

aidanhs replied:
TestBuildAddNonTar should pass both on master and on the branch. The test is to ensure that the current behaviour of adding a totally invalid tar (i.e. one which does not start with a tar header) is preserved in future.

tiborvass replied:
Thanks for confirming

tiborvass commented:
LGTM, but I would like a confirmation for TestBuildAddNonTar before merging
